cloudbuild.yaml

```diff
@@ -6,8 +15,171 @@ substitutions:
   _MYSQL_ROOT_PASSWORD: ""
   _MYSQL_PASSWORD: ""
 options:
-  machineType: N1_HIGHCPU_8
+  machineType: E2_HIGHCPU_32
+  env:
+    - GO111MODULE=on
+    - GOPROXY=https://proxy.golang.org
+    - PROJECT_ROOT=github.com/google/trillian
+    - GOPATH=/go
+    - DOCKER_CLIENT_TIMEOUT=120
+    - COMPOSE_HTTP_TIMEOUT=120
+
 steps:
+# First build a "trillian_testbase" docker image which contains most of the tools we need for the later steps:
+- name: 'gcr.io/cloud-builders/docker'
+  entrypoint: 'bash'
+  args: ['-c', 'docker pull gcr.io/$PROJECT_ID/trillian_testbase:latest || exit 0']
+- name: 'gcr.io/cloud-builders/docker'
+  args: [
+    'build',
+    '-t', 'gcr.io/$PROJECT_ID/trillian_testbase:latest',
+    '--cache-from', 'gcr.io/$PROJECT_ID/trillian_testbase:latest',
+    '-f', './integration/cloudbuild/testbase/Dockerfile',
+    '.'
+  ]
+
+# Set up tools and any other common steps which should not be part of Docker image.
+- id: prepare
+  name: 'gcr.io/${PROJECT_ID}/trillian_testbase'
+  entrypoint: ./integration/cloudbuild/prepare.sh
+
+# Run lint and porcelain checks
+- id: lint
+  name: 'gcr.io/${PROJECT_ID}/trillian_testbase'
+  entrypoint: bash
+  args:
+    - -exc
+    - |
+      curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.21.0
+      golangci-lint run --deadline=8m
+      ./scripts/check_license.sh $(find . -name '*.go' | grep -v mock_ | grep -v .pb.go | grep -v .pb.gw.go | grep -v _string.go | tr '\n' ' ')
+  waitFor:
+    - prepare
+
+# Run Bazel check
+- id: bazel
+  name: 'gcr.io/${PROJECT_ID}/trillian_testbase'
+  entrypoint: bash
+  args:
+    - -exc
+    - |
+      bazel build //:*
+  waitFor:
+    - prepare
+
+# TODO(al): This needs fixing
+# # Run Docker check
+# - id: docker
+#   name: 'gcr.io/${PROJECT_ID}/trillian_testbase'
+#   entrypoint: ./integration/docker_compose_integration_test.sh
+#   waitFor:
+#     - prepare
+
+# Presubmit
+- id: presubmit
+  name: 'gcr.io/${PROJECT_ID}/trillian_testbase'
+  entrypoint: ./integration/cloudbuild/run_presubmit.sh
+  args:
+    - --no-linters
+    - --no-generate
+  env:
+    - GOFLAGS=-race
+    - GOPATH=/throwaway
+    - GO_TEST_TIMEOUT=20m
+  waitFor:
+    - prepare
+
+# Presubmit (Batched queue)
+- id: presubmit_batched
+  name: 'gcr.io/${PROJECT_ID}/trillian_testbase'
+  entrypoint: ./integration/cloudbuild/run_presubmit.sh
+  args:
+    - --no-linters
+    - --no-generate
+  env:
+    - GOFLAGS=-race --tags=batched_queue
+    - GOPATH=/throwaway
+    - GO_TEST_TIMEOUT=20m
+  waitFor:
+    - presubmit
+
+# Presubmit (PKCS11)
+- id: presubmit_pkcs11
+  name: 'gcr.io/${PROJECT_ID}/trillian_testbase'
+  entrypoint: ./integration/cloudbuild/run_presubmit.sh
+  args:
+    - --no-linters
+    - --no-generate
+  env:
+    - GOFLAGS=-race --tags=pkcs11
+    - GO_TEST_TIMEOUT=20m
+  waitFor:
+    - presubmit
```

Comment:
There is a lot going on here and below. Should we attempt to add one build at a time, for easier reviewability?